packages/element/src/create-interpolate-element.js

```diff
@@ -0,0 +1,330 @@
+ * For example, for the given string:
+ *
+ * "This is a <span>string</span> with <a>a link</a> and a self-closing
+ * <CustomComponentB/> tag"
+ *
+ * You would have something like this as the conversionMap value:
+ *
+ * ```js
+ * {
+ *     span: <span />,
+ *     a: <a href={ 'https://github.com' } />,
+ *     CustomComponentB: <CustomComponent />,
+ * }
+ * ```
+ *
+ * @param {string}  interpolatedString  The interpolation string to be parsed.
+ * @param {Object}  conversionMap       The map used to convert the string to
+ *                                      a react element.
+ * @throws {TypeError}
+ * @return {Element}  A react element.
+ */
+const createInterpolateElement = ( interpolatedString, conversionMap ) => {
+	indoc = interpolatedString;
+	offset = 0;
+	output = [];
+	stack = [];
+	tokenizer.lastIndex = 0;
+
+	if ( ! isValidConversionMap( conversionMap ) ) {
+		throw new TypeError(
+			`The conversionMap provided is not valid. It must be an object with values that are WPElements`
+		);
+	}
+
+	do {
+		// twiddle our thumbs
+	} while ( proceed( conversionMap ) );
+
+	return ! output.every( ( element ) => typeof element === 'string' ) ?
+		createElement( Fragment, {}, output ) :
+		output.join( '' );
+};
+
+/**
+ * Validate conversion map.
+ *
+ * A map is considered valid if it's an object and every value in the object
+ * is a WPElement
+ *
+ * @private
+ * @param {Object} conversionMap  The map being validated.
+ *
+ * @return {boolean}  True means the map is valid.
+ */
+const isValidConversionMap = ( conversionMap ) => {
+	const isObject = typeof conversionMap === 'object';
+	const values = isObject && Object.values( conversionMap );
+	return isObject &&
+		values.length &&
+		values.every( ( element ) => isValidElement( element ) );
+};
+
+/**
+ * This is the iterator over the matches in the string.
+ *
+ * @private
+ * @param {Object} conversionMap The conversion map for the string.
+ *
+ * @return {boolean} true for continuing to iterate, false for finished.
+ */
+function proceed( conversionMap ) {
+	const next = nextToken();
+	const [ tokenType, name, startOffset, tokenLength ] = next;
+	const stackDepth = stack.length;
+	const leadingTextStart = startOffset > offset ? offset : null;
+	if ( ! conversionMap[ name ] ) {
+		if ( stackDepth !== 0 ) {
+			const { stackLeadingText, tokenStart } = stack.pop();
+			output.push( indoc.substr( stackLeadingText, tokenStart ) );
+		}
+		addText();
+		return false;
+	}
+	switch ( tokenType ) {
+		case 'no-more-tokens':
+			if ( stackDepth !== 0 ) {
+				const { stackLeadingText, tokenStart } = stack.pop();
+				output.push( indoc.substr( stackLeadingText, tokenStart ) );
+			}
+			addText();
+			return false;
+
+		case 'self-closed':
+			if ( 0 === stackDepth ) {
+				if ( null !== leadingTextStart ) {
+					output.push(
+						indoc.substr( leadingTextStart, startOffset - leadingTextStart )
+					);
+				}
+				output.push( cloneElement( conversionMap[ name ], { key: name } ) );
```

Comment:
We're not just passing in the name or type of element to create though - we are also passing in a real component with may have props on it already…

```js
interpolateComponents(
	'<Icon /> Do it!', {
	Icon: <Gridicon name="star" />
} )
```

so I think if we do `createElement` won't it lose the `star` attribute?

Reply:
My suggestion is:

```suggestion
				output.push( conversionMap[ name ] );
```

If the change is made, tests fail for two reasons:

- Snapshots. Easily updated.
- Because React warns about keys. But if we change the return value in `createInterpolateElement` to create the element using `...output` instead of `output`, this resolves the warning, because it's no longer treated as an arbitrary array of elements.

Reply:
It's unclear to me how your suggested change will handle balanced tags?  `cloneElement` here handles making sure the nested content inside balanced tags is added to the right element.

Reply:
Here's a diff with passing tests. Can you tell me what I've done wrong, or what tests need to be added to demonstrate the failure you're describing?

```diff
diff --git a/packages/element/src/create-interpolate-element.js b/packages/element/src/create-interpolate-element.js
index 7663dce24..ca7abf9c6 100644
--- a/packages/element/src/create-interpolate-element.js
+++ b/packages/element/src/create-interpolate-element.js
@@ -127,9 +127,7 @@ const createInterpolateElement = ( interpolatedString, conversionMap ) => {
 		// twiddle our thumbs
 	} while ( proceed( conversionMap ) );
 
-	return ! output.every( ( element ) => typeof element === 'string' ) ?
-		createElement( Fragment, {}, output ) :
-		output.join( '' );
+	return createElement( Fragment, null, ...output );
 };
 
 /**
@@ -188,7 +186,7 @@ function proceed( conversionMap ) {
 						indoc.substr( leadingTextStart, startOffset - leadingTextStart )
 					);
 				}
-				output.push( cloneElement( conversionMap[ name ], { key: name } ) );
+				output.push( conversionMap[ name ] );
 				offset = startOffset + tokenLength;
 				return true;
 			}
@@ -298,7 +296,7 @@ function addChild( component, tokenStart, tokenLength, lastOffset ) {
 	}
 
 	parent.component.children.push(
-		cloneElement( component.element, { key: component.name }, component.children )
+		cloneElement( component.element, null, ...component.children )
 	);
 	parent.prevOffset = lastOffset ? lastOffset : tokenStart + tokenLength;
 }
@@ -321,8 +319,8 @@ function addComponentFromStack( endOffset ) {
 	output.push(
 		cloneElement(
 			component.element,
-			{ key: component.name },
-			component.children
+			null,
+			...component.children
 		)
 	);
 }
diff --git a/packages/element/src/test/create-interpolate-element.js b/packages/element/src/test/create-interpolate-element.js
index 274ba94fc..2991729cf 100644
--- a/packages/element/src/test/create-interpolate-element.js
+++ b/packages/element/src/test/create-interpolate-element.js
@@ -18,12 +18,13 @@ describe( 'createInterpolateElement', () => {
 	} );
 	it( 'returns same string when there are no tokens in the string', () => {
 		const testString = 'This is a string';
+		const expectedElement = <>{ testString }</>;
 		expect(
 			createInterpolateElement(
 				testString,
 				{ someValue: <em /> }
 			)
-		).toBe( testString );
+		).toEqual( expectedElement );
 	} );
 	it( 'throws an error when there is an invalid conversion map', () => {
 		const testString = 'This is a <someValue/> string';
@@ -50,45 +51,46 @@ describe( 'createInterpolateElement', () => {
 	it( 'returns same string when there is an non matching token in the ' +
      'string', () => {
 		const testString = 'This is a <non_parsed/> string';
+		const expectedElement = <>{ testString }</>;
 		expect(
 			createInterpolateElement(
 				testString,
 				{ someValue: <strong /> }
 			)
-		).toBe( testString );
+		).toEqual( expectedElement );
 	} );
 	it( 'returns same string when there is spaces in the token', () => {
 		const testString = 'This is a <spaced token/>string';
+		const expectedElement = <>{ testString }</>;
 		expect(
 			createInterpolateElement(
 				testString,
 				{ 'spaced token': <em /> }
 			)
-		).toBe( testString );
+		).toEqual( expectedElement );
 	} );
 	it( 'returns string without unbalanced tags', () => {
 		const testString = 'This is a <em>string';
+		const expectedElement = <>{ 'This is a ' }{ 'string' }</>;
 		expect(
 			createInterpolateElement(
 				testString,
 				{ em: <em /> }
 			)
-		).toBe( 'This is a string' );
+		).toEqual( expectedElement );
 	} );
 	it( 'returns expected react element for non nested components', () => {
 		const testString = 'This is a string with <a>a link</a>.';
 		const expectedElement = createElement(
 			Fragment,
 			{},
-			[
-				'This is a string with ',
-				createElement(
-					'a',
-					{ href: 'https://github.com', className: 'some_class', key: 'a' },
-					[ 'a link' ]
-				),
-				'.',
-			]
+			'This is a string with ',
+			createElement(
+				'a',
+				{ href: 'https://github.com', className: 'some_class' },
+				'a link'
+			),
+			'.',
 		);
 		const component = createInterpolateElement(
 			testString,
@@ -108,22 +110,18 @@ describe( 'createInterpolateElement', () => {
 		const expectedElement = createElement(
 			Fragment,
 			{},
-			[
-				'This is a ',
+			'This is a ',
+			createElement(
+				'a',
+				null,
+				'string that is ',
 				createElement(
-					'a',
-					{ key: 'a' },
-					[
-						'string that is ',
-						createElement(
-							'em',
-							{ key: 'em' },
-							[ 'linked' ]
-						),
-					]
+					'em',
+					null,
+					'linked'
 				),
-				'.',
-			]
+			),
+			'.',
 		);
 		expect( JSON.stringify( createInterpolateElement(
 			testString,
@@ -142,14 +140,12 @@ describe( 'createInterpolateElement', () => {
 		const expectedElement = createElement(
 			Fragment,
 			{},
-			[
-				'This is a string with a ',
-				createElement(
-					TestComponent,
-					{ key: 'TestComponent' },
-					[ 'Custom Component' ]
-				),
-			]
+			'This is a string with a ',
+			createElement(
+				TestComponent,
+				null,
+				'Custom Component'
+			),
 		);
 		expect( JSON.stringify( createInterpolateElement(
 			testString,
@@ -166,13 +162,10 @@ describe( 'createInterpolateElement', () => {
 		const expectedElement = createElement(
 			Fragment,
 			{},
-			[
-				'This is a string with a self closing custom component: ',
-				createElement(
-					TestComponent,
-					{ key: 'TestComponent' }
-				),
-			]
+			'This is a string with a self closing custom component: ',
+			createElement(
+				TestComponent
+			),
 		);
 		expect( JSON.stringify( createInterpolateElement(
 			testString,
@@ -198,27 +191,22 @@ describe( 'createInterpolateElement', () => {
 		const expectedElement = createElement(
 			Fragment,
 			{},
-			[
-				'This is a complex string with a ',
-				createElement(
-					'a',
-					{ key: 'a1' },
-					[
-						'nested ',
-						createElement(
-							'em',
-							{ key: 'em1' },
-							[ 'emphasized string' ]
-						),
-						' link',
-					]
-				),
-				' and value: ',
+			'This is a complex string with a ',
+			createElement(
+				'a',
+				null,
+				'nested ',
 				createElement(
-					TestComponent,
-					{ key: 'TestComponent' }
+					'em',
+					null,
+					'emphasized string'
 				),
-			]
+				' link',
+			),
+			' and value: ',
+			createElement(
+				TestComponent
+			),
 		);
 		expect( JSON.stringify( createInterpolateElement(
 			testString,
@@ -253,17 +241,13 @@ describe( 'createInterpolateElement', () => {
 		const expectedElement = createElement(
 			Fragment,
 			{},
-			[
-				'👳‍♀️',
-				createElement(
-					'strong',
-					{ key: 'icon' },
-					[
-						'🚨🤷‍♂️⛈️fully',
-					]
-				),
-				' here',
-			]
+			'👳‍♀️',
+			createElement(
+				'strong',
+				null,
+				'🚨🤷‍♂️⛈️fully',
+			),
+			' here',
 		);
 		expect( JSON.stringify( createInterpolateElement(
 			testString,
```

Reply:
Oops I confused which `cloneElement` you were modifying 🤦‍♂

Reply:
I definitely like removing the keys here 👍 I'll make these changes.

Reply:
see ec9afec - love how this simplifies things, also a TIL on the `createElement` signature.  Thanks for that (although I probably should have noticed that myself from the docs :| )